include/pybind11/stl.h

```diff
@@ -305,6 +316,55 @@ template<> struct type_caster<std::experimental::nullopt_t>
     : public void_caster<std::experimental::nullopt_t> {};
 #endif
 
+#if defined(PYBIND11_HAS_FILESYSTEM)
+template<> struct type_caster<std::filesystem::path> {
+
+    static handle cast(const std::filesystem::path& path, return_value_policy, handle) {
+        PyObject* py_str = [](const auto& w) {
+            using char_type = std::filesystem::path::value_type;
+            if constexpr (std::is_same_v<char_type, char>) {
+                return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
+            } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+                return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
+            }
+        } (path.native());
+        if (!py_str) {
+            return nullptr;
+        }
+        return module::import("pathlib").attr("Path")(reinterpret_steal<object>(py_str))
+               .release();
+    }
+
+    bool load(handle handle, bool) {
+        using char_type = std::filesystem::path::value_type;
+        PyObject* tmp = nullptr;
+        if constexpr (std::is_same_v<char_type, char>) {
+            if (PyUnicode_FSConverter(handle.ptr(), &tmp)) {
+                if (auto c_str = PyBytes_AsString(tmp)) {
+                    value = c_str;
+                }
+            }
+        } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+            if (PyUnicode_FSDecoder(handle.ptr(), &tmp)) {
+                if (auto c_str = PyUnicode_AsWideCharString(tmp, nullptr)) {
+                    value = c_str;
+                    PyMem_Free(c_str);
+                }
+            }
+        }
+        Py_XDECREF(tmp);
+        if (PyErr_Occurred()) {
+            PyErr_Clear();
+            return false;
```

Comment:
It might be a good idea to move these clears and returns closer to the location where you could return. E.g., after `value = c_str;`, you could `return true;`.

Reason for this: first of all, it makes the decision clearer at the spot where it's tested. Second reason to also move up `PyErr_Clear`: you cannot call other C API functions with this error set.

Both would follow the style from the other casters more closely.

Reply:
Oh, but what about `Py_XDECREF(tmp);`? :-/

Can we somehow turn `tmp` into a `py::object`? Or otherwise, to clarify the flow in `load`, it might be nice to move the `PyUnicode_FSDecoder` to a private (maybe static) function. Other casters also often do this.

Reply:
Just a nudge/suggestion, though. Maybe you tried this, and there are good reasons for the current way things look!

Reply:
I'm not really a fan of using `py::object` here given that the whole point is to interact with the CPython C-API (via PyUnicode_\* and PyBytes_\*), so we'd have to keep wrapping and unwrapping things...  In "classic C" style having all the cleanup at the end (perhaps even using `goto exit;` on errors) is pretty standard, so at least I'm more comfortable using that.  It's not even clear to me how exactly py::object would handle refcounts wrt. being set a new value in `PyUnicode_FSConverter(..., &tmp)`.

To be clear I'm not opposed to using py::object here... I'm just writing in the style I'm comfortable with, but if you want to propose your own patch here I'm fine too.

Reply:
Fair enough. Just an idea to see if this could work/be more homogeneous with other casters. As long as the code doesn't get more complex this is fine. I guess I'm just more comfortable with C++, where you try to make sure you don't forget things with RAII and be able to return in the middle.

>  It's not even clear to me how exactly py::object would handle refcounts wrt. being set a new value in `PyUnicode_FSConverter(..., &tmp)`.

True, yes. That's an unusual interface for the Python C API.

Reply:
So, to summarize, unless @bstaletic/@henryiii/other reviewers think this can be clean up/made more similar to other casters, I'm fine with the current situation, and we can just see how this piece of code evolves.

Reply:
With a bunch of `.ptr()` sprinkled around, I'm not sure `py::object` would be cleaner.